### ISSUE TITLE
Incorrect and inconsistent behaviour using HashSet::into_iter

### ISSUE THREAD

**Issue by andrewhickman** [bodil/im-rs#60]
The following code occasionally fails:
```
    for i in 0..10000 {
        let mut lhs = vec![604, 986, 436, 379, 59, 423, 391, 434, 375, 486, 64, 851, 631, 89];
        lhs.sort();

        let mut iset = im::HashSet::new();
        for &i in &lhs {
            iset.insert(i);
        }

        let mut rhs: Vec<_> = iset.clone().into_iter().collect();
        rhs.sort();

        if lhs != rhs {
            println!("iteration: {}", i);
            println!("lhs: {:?}", &lhs);
            println!("rhs: {:?}", &rhs);
            panic!();
        }
    }
```
It seems values are sometimes skipped in the IntoInterator implementation of HashSet. 

The behaviour is not consistent but 10000 iterations is usually enough to reproduce it, outputting something like
```
iteration: 3370
lhs: [59, 64, 89, 375, 379, 391, 423, 434, 436, 486, 604, 631, 851, 986]
rhs: [64, 375]
```

**Comment by bodil** [comment]
Likely a bug in the unsafe chunk code which fails to initialise some allocated memory before use. This should be a fun one to track down...

**Comment by bodil** [comment]
The `mem:zeroed()` patch doesn't seem to solve the issue, and I can't for the life of me find anything wrong with the implementation of `SparseChunk`, so I'm leaning towards a bug in hash collision handling, which I think must be the only other thing that could be nondeterministic in this way.

**Comment by Eh2406** [comment]
If this was reading undefined memory, should the problematic versions be yanked and maybe a [security advisory be issued](https://rustsec.org/advisories/)?

**Comment by bodil** [comment]
I don't have any reason to believe it was reading uninitialised memory in any meaningful way, though I also didn't really find a satisfying answer to what was going on. The bug was manifesting in code from `std`, but may have been due to some weird interaction with some of the unsafe code I removed when fixing it.

I'm still not disinclined to yank all minor versions previous to the fix, though... I think I'll go ahead and do that.
